src/main/java/org/mockito/internal/creation/bytebuddy/CachingMockBytecodeGenerator.java

```diff
@@ -1,70 +1,94 @@
 package org.mockito.internal.creation.bytebuddy;
 
-import static org.mockito.internal.util.StringJoiner.join;
+import org.mockito.exceptions.base.MockitoException;
+
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.SoftReference;
 import java.lang.ref.WeakReference;
 import java.lang.reflect.Modifier;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
-import java.util.WeakHashMap;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
-import org.mockito.exceptions.base.MockitoException;
+import java.util.concurrent.ConcurrentMap;
+
+import static org.mockito.internal.util.StringJoiner.join;
 
-class CachingMockBytecodeGenerator {
+class CachingMockBytecodeGenerator extends ReferenceQueue<ClassLoader> {
 
-    private final Lock avoidingClassLeakCacheLock = new ReentrantLock();
-    public final WeakHashMap<ClassLoader, CachedBytecodeGenerator> avoidingClassLeakageCache =
-            new WeakHashMap<ClassLoader, CachedBytecodeGenerator>();
+    private static final ClassLoader BOOT_LOADER = new URLClassLoader(new URL[0], null);
+
+    final ConcurrentMap<Key<ClassLoader>, CachedBytecodeGenerator> avoidingClassLeakageCache =
+            new ConcurrentHashMap<Key<ClassLoader>, CachedBytecodeGenerator>();
 
     private final MockBytecodeGenerator mockBytecodeGenerator = new MockBytecodeGenerator();
 
+    private final boolean weak;
+
+    public CachingMockBytecodeGenerator(boolean weak) {
+        this.weak = weak;
+    }
+
     @SuppressWarnings("unchecked")
     public <T> Class<T> get(MockFeatures<T> params) {
-        // TODO improves locking behavior with ReentrantReadWriteLock ?
-        avoidingClassLeakCacheLock.lock();
-        try {
-
-            Class<?> generatedMockClass = mockCachePerClassLoaderOf(params.mockedType).getOrGenerateMockClass(
-                    params
-            );
+        pollAndClearReferences();
+        return (Class<T>) mockCachePerClassLoaderOf(params.mockedType.getClassLoader()).getOrGenerateMockClass(params);
+    }
 
-            return (Class<T>) generatedMockClass;
-        } finally {
-          avoidingClassLeakCacheLock.unlock();
+    void pollAndClearReferences() {
```

Comment:
I would like a comment on this, or maybe rename it better (I'm thinking loudly).
Not everyone knows `ReferenceQueue` or how it works, and thus why we need to invoke this method upon `get(...)`.

Ideas welcome.